compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -635,7 +635,26 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 self.codegen_consume(bx, place.as_ref())
             }
 
-            mir::Operand::Constant(ref constant) => self.eval_mir_constant_to_operand(bx, constant),
+            mir::Operand::Constant(ref constant) => {
+                let constant_ty = constant.ty();
+                // Make sure that constant SIMD vectors are passed as immediates.
+                // This is needed for some intrinsics.
+                return if constant_ty.is_simd() {
```

Comment:
Is the is_simd() check still needed?

Reply:
When I compiled the compiler I got the following error: `error: failed to get layout for Option<&mut T>: the type `<T as ptr::metadata::Pointee>::Metadata` has an unknown layout`.

Reply:
That sounds like you need to normalize the type first: `tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), ty)`.

`is_simd` should not be needed.

Reply:
Ah never mind, we do need `is_simd` or else the `simd_size_and_type` later will go wrong.

But if you got an un-normalized type, as it seems from that error, then normalization should also be added.

Reply:
To clarify here, does this check need changing? Or is it sufficient as is?

Reply:
You should add a call to `normalize_erasing_regions` before `ty.is_simd()` since apparently we cannot rely on the input type being normalized.

Reply:
Calling that function causes the following error:
```
error: internal compiler error: 
compiler/rustc_middle/src/ty/normalize_erasing_regions.rs:169:90: Failed to normalize
 FnDef(DefId(2:3484 ~ core[23d2]::ops::bit::Not::not), 
 [Alias(Projection, AliasTy { args: [F/#0], def_id: DefId(0:1958 ~ compiler_builtins[2e97]::float::Float::Int), .. })]), 
 maybe try to call `try_normalize_erasing_regions` instead
```

Taking the advice of that error how does following seem?:
```rs
let constant_ty_result =
    bx.tcx().try_normalize_erasing_regions(ty::ParamEnv::reveal_all(), constant.ty());
// Make sure this type is actually passed as an immediate vector.
// (In particular, this excludes packed SIMD vectors.)
if let Ok(constant_ty) = constant_ty_result && constant_ty.is_simd() {
    let layout = bx.layout_of(constant_ty);
    
    // code as it currently is
}
```

Reply:
No, the `try` version should not be needed during codegen.

Not quite sure what is happening here... Cc @lcnr we're having trouble with normalization.^^

Reply:
Can you try `self.monomorphize(constant.ty())`? That is what `immediate_const_vector` also calls.

Reply:
Yes, calling `self.monomorphize(constant.ty())` works fine:

```rs
let constant_ty = self.monomorphize(constant.ty());
// Make sure this type is actually passed as an immediate vector.
// (In particular, this excludes packed SIMD vectors.)
if constant_ty.is_simd() {
    let layout = bx.layout_of(constant_ty);

    // code as it currently is
}
```

Reply:
That looks good to me now. :)